Make cloneObject work in both Node & browsers

paambaati opened the pull request:
This fixes #3570 by ensuring we check for `File` to be available as a global before we do an instance check. This is required as `File` isn't available on Node.js environments.

@bluebill1049 I'm not sure how to test for this, so any help there is appreciated.

paambaati committed:
Make cloneObject work in both Node & browsers

This fix ensures we check for `File` to be available as a global before we do an instance check. This is required as `File` isn't available on Node.js environments.

paambaati committed:
Make cloneObject work in both Node & browsers

This fix ensures we check for `File` to be available as a global before we do an instance check. This is required as `File` isn't available on Node.js environments.

paambaati committed:
Try to fix lint errors.

paambaati committed:
Round #2 - Attempt to fix lint errors.

paambaati commented:
@bluebill1049 On closer analysis, perhaps adding the `isWeb` parameter to the `cloneObject()` call here too would be a more robust fix? Thoughts?

https://github.com/react-hook-form/react-hook-form/blob/4407116f36df21de2247b839802028aa8cb0efe5/src/useFieldArray.ts#L152

bluebill1049 commented on src/utils/cloneObject.ts:
```diff
@@ -6,7 +6,10 @@ export default function cloneObject<T extends unknown>(
-  if (isPrimitive(data) || (isWeb && data instanceof File)) {
+  if (
+    isPrimitive(data) ||
+    (isWeb && typeof File === 'function' && data instanceof File)
```
I think the issue is probably caused due to isWeb is not passing down to the function.

shall we just keep `typeof window === UNDEFINED` inline and remove that extra isWeb argument?

bluebill1049 replied:
actually, I think `isWindowUndefined` will do.

bluebill1049 replied:
`const isWindowUndefined = typeof window === UNDEFINED;`

paambaati replied:
@bluebill1049 Yeah using `isWindowUndefined` seems good enough. Why was `isWeb` introduced in the first place? My initial guess was that it made it easier to test/mock browsers & Node in tests.

bluebill1049 replied:
yes.

paambaati replied:
@bluebill1049 Done, just pushed. I decided to call it `isWeb` (instead of using `const isWindowUndefined = typeof window === UNDEFINED;` like you suggested) in `cloneObject.ts` because it made its purpose more explicit.

bluebill1049 replied:
hmmm looks like there is react native as well to take care.

paambaati replied:
Not sure how `window` globals work in RN; should we then go back to checking if `File` exists? Like `typeof File === 'function'`?

bluebill1049 replied:
hmmm, let me do some quick testing.

bluebill1049 commented:
Thanks, heaps for the PR 🙏

paambaati committed:
clone-object: A better for the cloneObject() method.

Remove the isWeb parameter entirely and instead check it right inside the implementation.

bluebill1049 commented on src/utils/cloneObject.ts:
```diff
@@ -1,9 +1,9 @@
-  data: T,
-  isWeb = true,
-): T {
+const isWeb = typeof window !== UNDEFINED;
```
```suggestion
const isWeb = typeof window !== UNDEFINED && !isUndefined(HTMLElement);
```

bluebill1049 replied:
see if this fix react-native tests.

paambaati committed:
Update src/utils/cloneObject.ts

Include @bluebill1049's suggested fix.

Co-authored-by: Bill <bluebill1049@hotmail.com>

paambaati committed:
cloneObject - Add missing import.

bluebill1049 commented:
could you please try this beta? `6.12.2-beta.1`

bluebill1049 committed:
move isWeb logic into its own file

paambaati commented:
@bluebill1049 Can confirm that `6.12.2-beta.1` works! 🥳

bluebill1049 commented:
> @bluebill1049 Can confirm that `6.12.2-beta.1` works! 🥳

Thank you so much for your help!
